Mathlib/NumberTheory/EllipticDivisibilitySequence.lean

```diff
@@ -65,26 +74,500 @@ elliptic, divisibility, sequence
 
 universe u v w
 
-variable {R : Type u} {S : Type v} [CommRing R] [CommRing S] (W : ℤ → R) (f : R →+* S)
+variable {R : Type u} {S : Type v} [CommRing R] [CommRing S] (W : ℤ → R)
+variable {F} [FunLike F R S] [RingHomClass F R S] (f : F)
+
+open scoped nonZeroDivisors
+
+namespace EllSequence
+
+/-- The expression `W((m+n)/2) * W((m-n)/2)` is the basic building block of elliptic relations,
+where integers `m` and `n` should have the same parity. -/
+def addMulSub (m n : ℤ) : R := W ((m + n).div 2) * W ((m - n).div 2)
+/- Implementation note: we use `Int.div _ 2` instead of `_ / 2` so that `(-m).div 2 = -(m.div 2)`
+and lemmas like `addMulSub_neg₀` hold unconditionally, even though in the case we care about
+(`m` and `n` both even or both odd) both are equal. -/
+
+/-- The four-index elliptic relation, defined in terms of `addMulSub`,
+featuring the three partitions of four indices into two pairs.
+Intended to apply to four integers of the same parity. -/
+def rel₄ (a b c d : ℤ) : R :=
+  addMulSub W a b * addMulSub W c d
+    - addMulSub W a c * addMulSub W b d
+    + addMulSub W a d * addMulSub W b c
+
+/-- The defining property of Stange's elliptic nets,
+equivalent to a suitable valid (same-parity indices) `rel₄` relation,
+but here only the first three indices enjoy symmetry under permutation,
+while in `rel₄` all four indices can be freely permuted.
+
+The order of the last two terms are changed and two signs are swapped compared to Stange's
+paper to make the equivalence with elliptic relations unconditional (indepedent of W
+being an odd function). This should also avoid peculiarities in characterstic 3. -/
+def net (p q r s : ℤ) : R :=
+  W (p + q + s) * W (p - q) * W (r + s) * W r
+    - W (p + r + s) * W (p - r) * W (q + s) * W q
+    + W (q + r + s) * W (q - r) * W (p + s) * W p
+
+variable {W} in
+lemma net_eq_rel₄ {p q r s : ℤ} :
+    net W p q r s = rel₄ W (2 * p + s) (2 * q + s) (2 * r + s) s := by
+  simp_rw [net, rel₄, addMulSub, add_add_add_comm _ s, add_sub_add_comm, sub_self, add_zero,
+    add_assoc, ← two_mul, add_sub_cancel_right, ← left_distrib, ← mul_sub_left_distrib,
+    Int.mul_div_cancel_left _ two_ne_zero]
+  ring
+
+/-- The three-index elliptic relation, obtained by
+specializing to `d = 0` in the four-index relation. -/
+def Rel₃ (m n r : ℤ) : Prop :=
```

Comment:
Here is my suggestion: make this into an term `rel₃` of `R` rather than a `Prop`, and replace instances of `Rel₃` elsewhere with `rel₃ .. = 0`. I prefer keeping this analogous to `rel₄` (especially if you decide to expose it, but also for people who might read the source code), and lemmas like `rel₃_iff_rel₄_eq_zero` suggest that `rel₃` and `rel₄` are terms of the same type. In any case, `W (m + n) * W (m - n) * W r ^ 2 - W (m + r) * W (m - r) * W n ^ 2 + W (n + r) * W (n - r) * W m ^ 2 = 0` might be more suggestive of its cyclic nature.

I still don't quite like `rel` because that suggests a `Prop`, but I guess it's a "relator" rather than a "relation".